### PR TITLE
[ATen] Keep the fill_ kernel for zero_ on ROCm 7.0.0-7.1.0 (hipMemsetAsync corrupts under graph capture) (#197017)

### PR DESCRIPTION
**Pull request by sreedal**
Summary:

`zero_` on CUDA was moved off the `fill_` kernel and onto `cudaMemsetAsync` in
#195303 (`df69a036b4c2`, "[ATen] Zero CUDA tensors with cudaMemsetAsync instead
of a fill kernel"), which added `zero_cuda_` and repointed the `CUDA` dispatch
key for `zero_` in `native_functions.yaml`. On affected ROCm versions that path
silently returns wrong data: `hipMemsetAsync` captured into a HIP graph leaves
stale bytes behind, so a buffer `zero_` reports as zeroed reads back non-zero.
No error, no NaN, correct shape and dtype, so nothing downstream notices.

This is a ROCm runtime regression, not a PyTorch one, and it is bounded:

| ROCm  | memset under graph capture |
|---|---|
| 6.4.2 | clean |
| 7.0.0 | broken |
| 7.1.0 | clean |

So `zero_cuda_` now takes the memset fast path everywhere except ROCm 7.0.0
through 7.1.0, which fall back to `self.fill_(0)` -- what `zero_` did before
#195303. CUDA is untouched. The exact ROCm release that fixed it has not been
pinned down, so the window closes at 7.2.4, the lowest version confirmed clean;
it can be tightened once that version is known.

Credit to pdesupinski for reducing this to `hipMemsetAsync` plus graph capture
and for establishing the version window. The original diff gated all of ROCm
because the mechanism was unclear at the time.

How it was found, for context on why the symptom looked stranger than the bug:
an LLM inference server on MI350X returned garbage completions. A decode step
zeroes a scratch buffer every iteration; when the zeroing does not take effect
the sampler reads stale logits and emits a long run of token id 0. The
production hit rate tracked concurrency -- 16% of responses at 64 concurrent
requests, 41% at 512, 6% at 8192 -- which is why light traffic looked clean.
Concurrency is not part of the trigger though; it only changes how often the
captured graph is replayed and how much stale data is sitting in the buffer.
Graph capture is the trigger, and disabling capture while leaving
`torch.compile` on was enough to make the corruption disappear:

| configuration | corrupt responses |
|---|---|
| memset, `torch.compile` on, graphs captured | 3.03% (174/5737) |
| memset, `torch.compile` on, graphs disabled | 0.00% (0/4656) |
| memset, `torch.compile` off, graphs disabled | 0.00% (0/5956) |
| `fill_` kernel, `torch.compile` on, graphs captured | 0.00% (0/5782) |

Authored with the assistance of an AI coding agent.

Test Plan:
Unit tests, `test/test_torch.py`:

* `test_zero_dense_emits_memset` asserts that dense `zero_` emits a `Memset`
  event. Which path `zero_` takes on ROCm is now version dependent, and the
  test-side ROCm version helper (`common_cuda.ROCM_VERSION`) is major/minor
  only, so it cannot express the 7.2.4 boundary. Rather than duplicate the
  version window in Python and have it rot against the C++, the test is skipped
  on ROCm and its original CUDA assertion is restored unchanged.
* `test_zero_dense_inside_graph_capture` is new, and is the one that actually
  covers this bug on ROCm. It captures `zero_` into a graph, replays it, and
  checks the buffer really is zero -- an assertion about behaviour rather than
  about which kernel ran, so it holds on every backend and every ROCm version.
  It is skipped where graph capture is unavailable.

```
python test/test_torch.py -k test_zero_
```

Honest status: these have not been run on a ROCm host. The MI350X capacity
reachable from here runs a packaged serving build, not a PyTorch source tree,
so `test_torch.py` cannot be invoked against it directly. The end-to-end runs
below exercise the same code path on the same hardware, but that is not a green
test run and is not claimed as one.

CUDA is unaffected by construction: `!defined(USE_ROCM)` makes the gate
unconditionally true on a CUDA build, so `zero_cuda_` compiles identically to
before this diff, and the profiler assertion on CUDA is the pre-existing one.

Minimal reproducer, pure PyTorch -- no serving stack, no model, no concurrency.
pdesupinski reduced the trigger to `hipMemsetAsync` under graph capture, so the
background load the earlier version of this test plan carried is not needed:

```python
import torch

buf = torch.empty(512, 151936, dtype=torch.float32, device="cuda")
sentinel = torch.full_like(buf, 7.0)
bad = torch.zeros((), dtype=torch.long, device="cuda")

def step():
    buf.copy_(sentinel)
    buf.zero_()
    bad.add_(buf.ne(0).sum())

side = torch.cuda.Stream()
side.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(side):
    for _ in range(3):
        step()
    graph = torch.cuda.CUDAGraph()
    with torch.cuda.graph(graph):
        step()
torch.cuda.current_stream().wait_stream(side)

bad.zero_()
for _ in range(20000):
    graph.replay()
torch.cuda.synchronize()
print("stale elements:", bad.item())   # expect 0
```

End-to-end evidence, collected before the mechanism was narrowed. A 0.6B
decoder-only model served with vLLM on one 8-GPU MI350X host, driven with a
fixed 200-token prompt and 100-token output, stop tokens disabled so every
request decodes the full 100 tokens. A response counts as corrupt if it
contains a run of token id 0. Each row is a separate run against a freshly
started server; checkpoint, sampling settings and every other flag held
constant, only the named column varied.

| concurrency | before (memset) | after (fill kernel) |
|---|---|---|
| 64 | 16.42% (33/201) | 0.00% (0/201) |
| 512 | 40.55% (339/836) | 0.00% (0/836) |
| 8192 | 6.31% (190/3012) | 0.00% (0/3000) |

Throughput and latency over the same runs, showing the fill kernel costs
nothing here:

| concurrency | | QPS | decode tok/s | avg latency | p50 TTFT | p99 TTFT |
|---|---|---|---|---|---|---|
| 64 | before | 89.09 | 8857 | 0.715s | 712.4ms | 824.1ms |
| 64 | after | 91.18 | 9041 | 0.699s | 700.4ms | 812.4ms |
| 512 | before | 561.44 | 55958 | 0.845s | 752.5ms | 2142.9ms |
| 512 | after | 566.42 | 56164 | 0.837s | 744.4ms | 2252.7ms |

`arc lint` on all four changed files: no issues.

Differential Revision: D119313733

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

File: aten/src/ATen/native/cuda/TensorFactories.cu [hunk 1]
```diff
@@ -34,6 +34,11 @@
 namespace at::native {
 
 Tensor& zero_cuda_(Tensor& self) {
+  // hipMemsetAsync silently leaves stale data behind when it is captured into a
+  // HIP graph, so affected ROCm versions use the fill_ kernel instead. Clean on
+  // 6.4.2, broken on 7.0.0, clean again on 7.2.4; the version that fixed it is
+  // not known, so the window closes at the lowest version confirmed clean.
+#if !defined(USE_ROCM) || ROCM_VERSION < 70000 || ROCM_VERSION >= 70204
```
Comment (pdesupinski) [inline aten/src/ATen/native/cuda/TensorFactories.cu]: See my comment on the diff -- please fix this to upper bound at 7.1.0.

**Review by pdesupinski — CHANGES_REQUESTED** [review]
I believe [this](https://github.com/ROCm/clr/commit/3038c4c3777eeb55eabfddee56807c6f74b1b77c?fbclid=IwY2xjawUVOWlwZG9mA2V4dG4DYWVtAjExAGJyaWQRMTlnSVJoY2lSWUh2QkcybERzcnRjBmFwcF9pZAEwAAEeqsmh02Ks0a1kmcLMMGH72wLqLh56_paXL3NdzMOSBXjtAkfPDcyR17KcJZU_aem_cfsooNWdQslQQ3iYWAzEYA) is the fix commit, and I have verified independently that this repros for 7.0.0 but not 7.1.0 or 6.4.2. So we should be good to bound this to between 7.0.0 and 7.1.0.

Let me know when you fix this, then let's get this landed. Ideally, you can also trim down the summary to be a little more human :), but I have enough context to review this anyway so it's not blocking.

**Review by pdesupinski — DISMISSED** [review]

**Comment by pdesupinski** [comment]
cc @ngimel for review

**Review by ngimel — DISMISSED** [review]
